description: implementation of template function must be in header to be exported
author: Michael Gilbert <mgilbert@debian.org>

--- a/v8/src/objects/js-objects.h
+++ b/v8/src/objects/js-objects.h
@@ -777,7 +777,23 @@ class JSObject : public TorqueGeneratedJ
   template <typename Dictionary>
   static void ApplyAttributesToDictionary(Isolate* isolate, ReadOnlyRoots roots,
                                           Handle<Dictionary> dictionary,
-                                          const PropertyAttributes attributes);
+                                          const PropertyAttributes attributes) {
+    int capacity = dictionary->Capacity();
+    for (int i = 0; i < capacity; i++) {
+      Object k;
+      if (!dictionary->ToKey(roots, i, &k)) continue;
+      if (k.FilterKey(ALL_PROPERTIES)) continue;
+      PropertyDetails details = dictionary->DetailsAt(i);
+      int attrs = attributes;
+      // READ_ONLY is an invalid attribute for JS setters/getters.
+      if ((attributes & READ_ONLY) && details.kind() == kAccessor) {
+        Object v = dictionary->ValueAt(i);
+        if (v.IsAccessorPair()) attrs &= ~READ_ONLY;
+      }
+      details = details.CopyAddAttributes(static_cast<PropertyAttributes>(attrs));
+      dictionary->DetailsAtPut(isolate, i, details);
+    }
+  }
 
  private:
   friend class JSReceiver;
